src/map/storage.c

```diff
@@ -212,21 +220,29 @@ int storage_additem(struct map_session_data* sd, struct item* item_data, int amo
  *------------------------------------------*/
 int storage_delitem(struct map_session_data* sd, int n, int amount)
 {
+	struct item *it = NULL;
+
 	nullpo_retr(1, sd);
-	Assert_retr(1, n >= 0 && n < MAX_STORAGE);
-	if( sd->status.storage.items[n].nameid == 0 || sd->status.storage.items[n].amount < amount )
+
+	Assert_retr(1, n >= 0 && n < VECTOR_LENGTH(sd->storage));
+
+	if (sd->state.storage_flag != STORAGE_FLAG_NORMAL)
+		return 0;
+
+	it = &VECTOR_INDEX(sd->storage, n);
+
+	if (it->nameid == 0 || it->amount < amount)
 		return 1;
 
-	sd->status.storage.items[n].amount -= amount;
-	if( sd->status.storage.items[n].amount == 0 )
-	{
-		memset(&sd->status.storage.items[n],0,sizeof(sd->status.storage.items[0]));
-		sd->status.storage.storage_amount--;
-		if( sd->state.storage_flag == STORAGE_FLAG_NORMAL )
-			clif->updatestorageamount(sd, sd->status.storage.storage_amount, MAX_STORAGE);
+	it->amount -= amount;
+
+	if (it->amount == 0) {
+		memset(it, 0, sizeof(struct item));
+		clif->updatestorageamount(sd, VECTOR_LENGTH(sd->storage), MAX_STORAGE);
```

Comment:
This isn't reporting the correct amount of items to the player (because by deleting an item in the middle, you can't rely on VECTOR_LENGTH() anymore. There needs to be a flag (in the sd) that counts the deleted items until the storage is closed and the vector is compacted...